DMTN-148.tex

```diff
@@ -50,9 +50,237 @@
 % ADD CONTENT HERE
 % You can also use the \input command to include several content files.
 
+\section{Intro}
+
+The goal of this tech note is to provide a plan for the handling of
+calibration products in the Gen-3 Butler environment.  There are a
+large number of these products currently in use, but many of them do
+not have a clearly defined process to define, ingest, test, and
+deprecate them.  The lack of documentation on how these processes work
+has led to many implementations of similar paths, further complicating
+the issue.
+
+
+\section{The End State}
+
+The final state that we wish to get to has the following properties:
+
+\begin{itemize}
+\item All calibration products are either lsst.afw.image.Exposure
+  image data or an explicit calibration class in the \verb|ip_isr| package.
+\item All calibration products are constructed from a DM Task in the
+  \verb|cp_pipe| package, and have the input data and configuration
+  parameters persisted with the calibration itself.  This guarantees
+  that any product can be reconstructed in the future.
+\item These provenance, config, and final product files may be saved
+  to an \verb|obs_CAMERA_data| package on git for convenience.  This also allows
+  these files to be migrated to other Butler repositories in a
+  consistent fashion.
+\item All calibration products are accepted for use and have a date
+  range set manually, preventing test runs from being used
+  inadvertently.  Calibrations ingested from git should supply
+  appropriate date ranges.
+\item Calibration information stored in a lsst.afw.cameraGeom object
+  may be used, but is always assumed to be overridden by supplied
+  calibration products.  No new calibration information should be
+  stored in these objects, and currently existing calibration
+  information should be deprecated.
+
+\end{itemize}
+
+\section{The Current State}
+
+\subsection{Standard Calibrations}
+
+The classic calibration products, \verb|BIAS|, \verb|DARK|, \verb|FLAT|, and \verb|FRINGE|, all
+write FITS files that are indexed in Butler's calibRegistry.sqlite3
+repository database.  These files are created by
+\verb|pipe_drivers|/constructCalibs.py, and are ingested into the
+repository upon acceptance by \verb|pipe_tasks|/ingestCalibs.py.  This
+ingestion process sets the 'validStart' and 'validEnd' values to
+restrict the usage of the calibration, as well as the 'calibDate',
+which is used (along with calibration type, filter, and detector) to
+uniquely identify the calibration on disk.
```

Comment:
When discussing `calibDate` with Tim, we realized that we need to set some ground rules for how this is set.  In reality it's an arbitrary string with the requirement that it is unique over all validity ranges for the smallest granularity for a calibration product (typically detectors).  Currently, the cp_pipe task that creates defects just uses the ISO string for the day.  This means that if all defects were reprocessed at once, all validity ranges would get the same `calibDate` and that would not be valid.

I think it would be good to try to go into some of that detail in this document since there is some subtlety to it.

Reply:
I agree.  I've changed the text to point that we currently use `calibDate`, but that the Butler collection cp_pipe used (for things we make that way) or the git SHA1 (for things we handle that way) make more sense.  I've punted on choosing a name for this, though.